File: backend-py/src/api/endpoints/sentry/handlers/issue_handler.py
from typing import Any, Mapping

from src import app
from src.models import Item, SentryInstallation, User
from src.models.items import ItemColumn
from src.database import db_session

from flask import Response


def handle_assigned(sentry_installation: SentryInstallation, issue_data: Mapping[str, Any]):
    # Find or create an item to associate with the Sentry Issue
    item, item_created = get_or_create_item(sentry_installation, issue_data)
    app.logger.info(f"{'Created' if item_created else 'Found'} linked Sentry issue")
    # Find or create a user to associate with the item
    # Note: The assignee in Sentry might be a team, which you could handle here as well
    assignee_data = issue_data.get('assignedTo', {})
    if assignee_data.get('type') == 'user':
        user, user_created = get_or_create_user(sentry_installation, assignee_data)
        item.assignee_id = user.id
        app.logger.info(
            f"Assigned to {'new' if user_created else 'existing'} user: {user.username}")
    db_session.commit()


def handle_created(sentry_installation: SentryInstallation, issue_data: Mapping[str, Any]):
    # Create an item to associate with the Sentry Issue
    item = Item(**get_item_defaults(sentry_installation, issue_data))
    db_session.add(item)
    db_session.commit()
    app.logger.info("Created linked Sentry issue")


def handle_ignored(sentry_installation: SentryInstallation, issue_data: Mapping[str, Any]):
    # Find or create an item to associate with the Sentry Issue
    item, item_created = get_or_create_item(sentry_installation, issue_data)
    app.logger.info(f"{'Created' if item_created else 'Found'} linked Sentry issue")
    # Mark the item as ignored
    item.is_ignored = True
    db_session.commit()
    app.logger.info("Marked item as ignored")


def handle_resolved(sentry_installation: SentryInstallation, issue_data: Mapping[str, Any]):
    # Find or create an item to associate with the Sentry Issue
    item, item_created = get_or_create_item(sentry_installation, issue_data)
    app.logger.info(f"{'Created' if item_created else 'Found'} linked Sentry issue")
    # Update the item's column to DONE
    item.column = ItemColumn.Done.value
    db_session.commit()
    app.logger.info(f"Updated item's column to {ItemColumn.Done.value}")


def issue_handler(action: str, sentry_installation: SentryInstallation, data: Mapping[str, Any]) -> Response:
    issue_data = data.get('issue')
    if action == "assigned":
        handle_assigned(sentry_installation, issue_data)
        return Response('', 202)
    elif action == "created":
        handle_created(sentry_installation, issue_data)
        return Response('', 201)
    elif action == "ignored":
        handle_ignored(sentry_installation, issue_data)
        return Response('', 202)
    elif action == 'resolved':
        handle_resolved(sentry_installation, issue_data)
        return Response('', 200)
    else:
        app.logger.info(f"Unhandled Sentry Issue action: {action}")
        return Response('', 400)


def get_item_defaults(sentry_installation: SentryInstallation, issue_data: Mapping[str, Any]):
    return {
        "organization_id": sentry_installation.organization_id,
        "title": issue_data.get('title'),
        "description": f"{issue_data.get('shortId')} - {issue_data.get('culprit')}",
        "column": ItemColumn.Done.value if issue_data.get('status') == "resolved" else ItemColumn.Todo.value,
        "is_ignored": issue_data.get('status') == "ignored",
        "sentry_id": issue_data.get('id'),
    }


def get_or_create_item(sentry_installation: SentryInstallation, issue_data: Mapping[str, Any]):
    item = Item.query.filter(
        Item.sentry_id == issue_data.get('id'),
        Item.organization_id == sentry_installation.organization_id
    ).first()
    if item:
        return item, False
    else:
        item = Item(**get_item_defaults(sentry_installation, issue_data))
        db_session.add(item)
        db_session.commit()
        return item, True


def get_or_create_user(sentry_installation: SentryInstallation, assignee_data: Mapping[str, Any]):
    user = User.query.filter(
        User.username == assignee_data.get('email'),
    ).first()
    if user:
        return user, False
    else:
        user = User(
            name=assignee_data.get('name'),
            username=assignee_data.get('email'),
            organization_id=sentry_installation.organization_id,
            avatar=f"https://ui-avatars.com/api/?name={assignee_data.get('name')}&background=random"
        )
        db_session.add(user)
        db_session.commit()
        return user, True

File: backend-ts/src/api/sentry/handlers/issueHandler.ts
import {Response} from 'express';

import Item, {ItemColumn} from '../../../models/Item.model';
import SentryInstallation from '../../../models/SentryInstallation.model';
import User from '../../../models/User.model';

async function handleAssigned(
  sentryInstallation: SentryInstallation,
  issueData: Record<string, any>
) {
  // Find or create an item to associate with the Sentry Issue
  const [item, isItemNew] = await Item.findOrCreate({
    where: {sentryId: issueData.id, organization_id: sentryInstallation.organizationId},
    defaults: {
      ...getItemDefaultsFromSentryIssue(sentryInstallation, issueData),
      column: ItemColumn.Doing,
    },
  });
  console.info(`${isItemNew ? 'Created' : 'Found'} linked item from Sentry issue`);
  // Find or create a user to associate with the item
  // Note: The assignee in Sentry might be a team, which you could handle here as well
  const {type, email, name} = issueData.assignedTo;
  if (type === 'user') {
    const [user, isUserNew] = await User.findOrCreate({
      where: {username: email},
      defaults: {
        name,
        username: email,
        avatar: `https://ui-avatars.com/api/?name=${name}&background=random`,
        organizationId: sentryInstallation.organizationId,
      },
    });
    await user.$add('items', item);
    console.info(`Assigned to ${isUserNew ? 'new' : 'existing'} user:`, user.username);
  }
}

async function handleCreated(
  sentryInstallation: SentryInstallation,
  issueData: Record<string, any>
) {
  // Create an item to associate with the Sentry Issue
  await Item.create(getItemDefaultsFromSentryIssue(sentryInstallation, issueData));
  console.info('Created linked item from Sentry issue');
}

async function handleIgnored(
  sentryInstallation: SentryInstallation,
  issueData: Record<string, any>
) {
  // Find or create an item to associate with the Sentry Issue
  const [item, isItemNew] = await Item.findOrCreate({
    where: {sentryId: issueData.id, organization_id: sentryInstallation.organizationId},
    defaults: getItemDefaultsFromSentryIssue(sentryInstallation, issueData),
  });
  console.info(`${isItemNew ? 'Created' : 'Found'} linked item from Sentry issue`);
  // Mark the item as ignored
  item.isIgnored = true;
  await item.save();
  console.info('Marked item as ignored');
}

async function handleResolved(
  sentryInstallation: SentryInstallation,
  issueData: Record<string, any>
) {
  // Find or create an item to associate with the Sentry Issue
  const [item, isItemNew] = await Item.findOrCreate({
    where: {sentryId: issueData.id, organization_id: sentryInstallation.organizationId},
    defaults: getItemDefaultsFromSentryIssue(sentryInstallation, issueData),
  });
  console.info(`${isItemNew ? 'Created' : 'Found'} linked item from Sentry Issue`);
  // Update the item's column to DONE
  item.column = ItemColumn.Done;
  await item.save();
  console.info(`Updated item's column to '${ItemColumn.Done}'`);
}

export default async function issueHandler(
  response: Response,
  action: string,
  sentryInstallation: SentryInstallation,
  data: Record<string, any>
): Promise<void> {
  const {issue: issueData} = data;

  switch (action) {
    case 'assigned':
      await handleAssigned(sentryInstallation, issueData);
      response.status(202);
      break;
    case 'created':
      await handleCreated(sentryInstallation, issueData);
      response.status(201);
      break;
    case 'ignored':
      await handleIgnored(sentryInstallation, issueData);
      response.status(202);
      break;
    case 'resolved':
      await handleResolved(sentryInstallation, issueData);
      response.status(202);
      break;
    default:
      console.info(`Unhandled Sentry Issue action: ${action}`);
      response.status(400);
  }
}

export const getItemDefaultsFromSentryIssue = (
  sentryInstallation: SentryInstallation,
  issueData: Record<string, any>
) => ({
  organizationId: sentryInstallation.organizationId,
  title: issueData.title as string,
  description: `${issueData.shortId} - ${issueData.culprit}`,
  column: issueData.status === 'resolved' ? ItemColumn.Done : ItemColumn.Todo,
  isIgnored: issueData.status === 'ignored',
  sentryId: issueData.id,
});
